test/functional/unified-spec-runner/match.ts

```diff
@@ -0,0 +1,301 @@
+import { expect } from 'chai';
+import { isDeepStrictEqual } from 'util';
+import { Binary, Document, Double, Int32, Long, ObjectId, Timestamp } from '../../../src';
+import {
+  CommandFailedEvent,
+  CommandStartedEvent,
+  CommandSucceededEvent
+} from '../../../src/cmap/events';
+import { CommandEvent, EntitiesMap } from './entities';
+import { ExpectedEvent } from './schema';
+
+export interface ExistsOperator {
+  $$exists: boolean;
+}
+export function isExistsOperator(value: unknown): value is ExistsOperator {
+  return typeof value === 'object' && value != null && '$$exists' in value;
+}
+export interface TypeOperator {
+  $$type: boolean;
+}
+export function isTypeOperator(value: unknown): value is TypeOperator {
+  return typeof value === 'object' && value != null && '$$type' in value;
+}
+export interface MatchesEntityOperator {
+  $$matchesEntity: string;
+}
+export function isMatchesEntityOperator(value: unknown): value is MatchesEntityOperator {
+  return typeof value === 'object' && value != null && '$$matchesEntity' in value;
+}
+export interface MatchesHexBytesOperator {
+  $$matchesHexBytes: string;
+}
+export function isMatchesHexBytesOperator(value: unknown): value is MatchesHexBytesOperator {
+  return typeof value === 'object' && value != null && '$$matchesHexBytes' in value;
+}
+export interface UnsetOrMatchesOperator {
+  $$unsetOrMatches: unknown;
+}
+export function isUnsetOrMatchesOperator(value: unknown): value is UnsetOrMatchesOperator {
+  return typeof value === 'object' && value != null && '$$unsetOrMatches' in value;
+}
+export interface SessionLsidOperator {
+  $$sessionLsid: string;
+}
+export function isSessionLsidOperator(value: unknown): value is SessionLsidOperator {
+  return typeof value === 'object' && value != null && '$$sessionLsid' in value;
+}
+
+export const SpecialOperatorKeys = [
+  '$$exists',
+  '$$type',
+  '$$matchesEntity',
+  '$$matchesHexBytes',
+  '$$unsetOrMatches',
+  '$$sessionLsid'
+];
+
+export type SpecialOperator =
+  | ExistsOperator
+  | TypeOperator
+  | MatchesEntityOperator
+  | MatchesHexBytesOperator
+  | UnsetOrMatchesOperator
+  | SessionLsidOperator;
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type KeysOfUnion<T> = T extends object ? keyof T : never;
+export type SpecialOperatorKey = KeysOfUnion<SpecialOperator>;
+export function isSpecialOperator(value: unknown): value is SpecialOperator {
+  return (
+    isExistsOperator(value) ||
+    isTypeOperator(value) ||
+    isMatchesEntityOperator(value) ||
+    isMatchesHexBytesOperator(value) ||
+    isUnsetOrMatchesOperator(value) ||
+    isSessionLsidOperator(value)
+  );
+}
+
+const TYPE_MAP = new Map();
+function typeof_double(actual): actual is number | Double {
```

Comment:
Just out of curiosity is there a pattern of using snake case in typescript? I hadn't seen this before.

Reply:
Could we perhaps have the functions inline in the Map.set calls? E.g.
```
TYPE_MAP.set('double', function (actual): actual is number | Double {
  return typeof actual === 'number' || actual._bsontype === 'Double';
});
```